File: src/process_data.py
import pandas as pd
import geopandas as gpd
from shapely.geometry import Point
import config
from pathlib import Path
from typing import Optional

def process_geography() -> gpd.GeoDataFrame:
    """
    Loads TIGER/Line shapefiles, filters for King County, calculates centroids,
    and assigns Treatment/Control status based on distance to transit stations.

    Returns:
        gpd.GeoDataFrame: A GeoDataFrame containing [GEOID, group, dist_to_station, geometry].
                          Rows in the 'Buffer' zone are excluded.
    """
    print("Processing Geography...")
    # Load Shapefile
    shp_path = config.DATA_RAW / "shapefiles" / f"tl_2020_53_bg.shp"
    if not shp_path.exists():
        raise FileNotFoundError(f"Shapefile not found at {shp_path}. Run download_data.py first.")
    
    gdf = gpd.read_file(shp_path)
    
    # Filter for King County (033)
    # GEOID format: State(2) + County(3) + Tract(6) + BlkGrp(1)
    # GEOID starts with '53033'
    gdf = gdf[gdf['GEOID'].str.startswith('53033')].copy()
    
    # Reproject to EPSG:3857 (Web Mercator - Meters) for distance calc
    gdf = gdf.to_crs(epsg=3857)
    
    # Calculate Centroids
    gdf['centroid'] = gdf.geometry.centroid
    
    # Define Stations
    stations_df = pd.DataFrame.from_dict(config.STATIONS, orient='index', columns=['lat', 'lon'])
    stations_gdf = gpd.GeoDataFrame(
        stations_df, 
        geometry=gpd.points_from_xy(stations_df.lon, stations_df.lat),
        crs="EPSG:4326"
    ).to_crs(epsg=3857)
    
    # Calculate Distances
    # Find distance to NEAREST station for each BG
    def get_min_dist(point):
        return stations_gdf.distance(point).min()
        
    gdf['dist_to_station'] = gdf['centroid'].apply(get_min_dist)
    
    # Classify
    def classify(dist: float) -> str:
        if dist <= config.TREATMENT_RADIUS_METERS:
            return 'Treatment'
        elif dist >= config.CONTROL_MIN_DIST_METERS:
            return 'Control'
        else:
            return 'Buffer'
            
    gdf['group'] = gdf['dist_to_station'].apply(classify)
    
    # Filter out Buffer
    gdf_analysis = gdf[gdf['group'] != 'Buffer'][['GEOID', 'group', 'dist_to_station', 'geometry']].copy()
    
    print(f"Geography ready. Treatment: {sum(gdf_analysis['group']=='Treatment')}, Control: {sum(gdf_analysis['group']=='Control')}")
    return gdf_analysis

def process_employment(geo_df: gpd.GeoDataFrame) -> pd.DataFrame:
    """
    Loads LODES WAC data for multiple years, aggregates to Block Group level,
    and merges with the geography dataframe.

    Args:
        geo_df (gpd.GeoDataFrame): The geography dataframe from process_geography().

    Returns:
        pd.DataFrame: A panel dataset ready for regression analysis.
    """
    print("Processing Employment Data...")
    dfs = []
    
    for year in config.YEARS:
        print(f"Loading {year}...")
        file_path = config.DATA_RAW / f"{config.STATE}_wac_S000_JT00_{year}.csv.gz"
        if not file_path.exists():
            print(f"Warning: {file_path} not found. Skipping.")
            continue
            
        # Read only necessary columns: w_geocode (Block ID), C000 (Total Jobs)
        # dtype={'w_geocode': str} is crucial to preserve leading zeros
        df = pd.read_csv(file_path, compression='gzip', usecols=['w_geocode', 'C000'], dtype={'w_geocode': str})
        
        # Create Block Group ID (First 12 chars)
        df['bg_id'] = df['w_geocode'].str.slice(0, 12)
        
        # Filter for King County (Optimization)
        df = df[df['bg_id'].str.startswith('53033')]
        
        # Aggregate to Block Group
        df_agg = df.groupby('bg_id')['C000'].sum().reset_index()
        df_agg.rename(columns={'C000': 'employment', 'bg_id': 'GEOID'}, inplace=True)
        
        # Merge with Geography
        merged = geo_df.merge(df_agg, on='GEOID', how='inner') # Inner join: keep only those with geo and data
        
        merged['year'] = year
        merged['post'] = int(year >= config.INTERVENTION_YEAR)
        merged['treated'] = (merged['group'] == 'Treatment').astype(int)
        
        dfs.append(merged)
        
    if not dfs:
        raise ValueError("No data processed.")
        
    final_df = pd.concat(dfs, ignore_index=True)
    return final_df

def main():
    geo_df = process_geography()
    panel_df = process_employment(geo_df)
    
    out_path = config.DATA_PROCESSED / "panel_dataset.csv"
    # Drop geometry for CSV saving
    panel_df.drop(columns='geometry').to_csv(out_path, index=False)
    print(f"Saved processed dataset to {out_path}")

if __name__ == "__main__":
    main()

File: src/process_data.R
# Process Data Script
library(sf)
library(dplyr)
library(readr)
library(stringr)

source("src/config.R")

process_geography <- function() {
  message("Processing Geography...")
  
  shp_path <- file.path(DATA_RAW, "shapefiles", "tl_2020_53_bg.shp")
  if (!file.exists(shp_path)) {
    stop("Shapefile not found. Run download_data.R first.")
  }
  
  # Load and filter for King County (FIPS 53033)
  gdf <- st_read(shp_path, quiet = TRUE) %>%
    filter(str_starts(GEOID, "53033")) %>%
    st_transform(crs = CRS_PROJECTED) # Project to WA North
  
  # Calculate Centroids
  # st_centroid returns a geometry column
  gdf$centroid <- st_centroid(gdf$geometry)
  
  # Define Stations
  stations_df <- data.frame(
    name = names(STATIONS),
    lat = sapply(STATIONS, function(x) x["lat"]),
    lon = sapply(STATIONS, function(x) x["lon"])
  )
  
  stations_sf <- st_as_sf(stations_df, coords = c("lon", "lat"), crs = 4326) %>%
    st_transform(crs = CRS_PROJECTED)
  
  # Calculate Distances
  # st_distance returns a matrix. We want min dist to ANY station for each BG.
  dist_matrix <- st_distance(gdf$centroid, stations_sf)
  gdf$dist_to_station <- apply(dist_matrix, 1, min)
  
  # Classify
  gdf <- gdf %>%
    mutate(group = case_when(
      dist_to_station <= TREATMENT_RADIUS_METERS ~ "Treatment",
      dist_to_station >= CONTROL_MIN_DIST_METERS ~ "Control",
      TRUE ~ "Buffer"
    ))
  
  # Filter out Buffer and select columns
  gdf_analysis <- gdf %>%
    filter(group != "Buffer") %>%
    select(GEOID, group, dist_to_station) # geometry is sticky in sf
  
  # Count groups
  counts <- table(gdf_analysis$group)
  message(sprintf("Geography ready. Treatment: %d, Control: %d", 
                  counts["Treatment"], counts["Control"]))
  
  return(gdf_analysis)
}

process_employment <- function(geo_df) {
  message("Processing Employment Data...")
  
  all_years_data <- list()
  
  for (year in YEARS) {
    message(sprintf("Loading %d...", year))
    file_path <- file.path(DATA_RAW, sprintf("%s_wac_S000_JT00_%d.csv.gz", STATE, year))
    
    if (!file.exists(file_path)) {
      warning(sprintf("%s not found. Skipping.", file_path))
      next
    }
    
    # Read CSV
    # col_types: w_geocode is character, C000 (Total Jobs) is double
    df <- read_csv(file_path, 
                   col_types = cols_only(w_geocode = col_character(), C000 = col_double()),
                   progress = FALSE)
    
    # Create Block Group ID (First 12 chars) and Filter for King County (53033)
    df_agg <- df %>%
      mutate(bg_id = str_sub(w_geocode, 1, 12)) %>%
      filter(str_starts(bg_id, "53033")) %>%
      group_by(bg_id) %>%
      summarise(employment = sum(C000, na.rm = TRUE)) %>%
      rename(GEOID = bg_id)
    
    # Join with Geography
    # Inner join to keep only matching records
    merged <- geo_df %>%
      inner_join(df_agg, by = "GEOID") %>%
      mutate(
        year = year,
        post = as.integer(year >= INTERVENTION_YEAR),
        treated = as.integer(group == "Treatment")
      )
    
    all_years_data[[as.character(year)]] <- merged
  }
  
  if (length(all_years_data) == 0) {
    stop("No data processed.")
  }
  
  final_df <- do.call(rbind, all_years_data)
  return(final_df)
}

process_data_main <- function() {
  geo_df <- process_geography()
  panel_df <- process_employment(geo_df)
  
  # Save to CSV (drop geometry for CSV)
  out_path <- file.path(DATA_PROCESSED, "panel_dataset.csv")
  st_set_geometry(panel_df, NULL) %>% # Drop geometry
    write_csv(out_path)
    
  message(sprintf("Saved processed dataset to %s", out_path))
}

if (sys.nframe() == 0) {
  process_data_main()
}
